WC shouldn't require ENABLE_WEBGL

donny-dont opened the pull request:
#### c4e0c5d7fd48af4a730c69902c84871d275b6d88
<pre>
WC shouldn&apos;t require ENABLE_WEBGL
<a href="https://bugs.webkit.org/show_bug.cgi?id=245188">https://bugs.webkit.org/show_bug.cgi?id=245188</a>

Reviewed by Fujii Hironori.

Make it so `USE(GRAPHICS_LAYER_WC)` can work without `ENABLE(WEBGL)`.
Guard `ENABLE(WEBGL)` usage and add a single `StreamConnectionWorkQueue`
to submit commands to. This is done due to thread safety issues with
ANGLE.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
* Source/WebKit/GPUProcess/graphics/wc/RemoteWCLayerTreeHost.cpp:
* Source/WebKit/UIProcess/playstation/PageClientImpl.cpp:
* Source/WebKit/UIProcess/playstation/PageClientImpl.h:
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:

Canonical link: <a href="https://commits.webkit.org/259825@main">https://commits.webkit.org/259825@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b330e51e67a3b7e2f2061659d005df6b5f424171

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105912 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14980 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38758 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115106 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16398 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6171 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98140 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114859 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111662 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12485 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95450 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40028 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94350 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27094 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81691 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8243 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28448 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8726 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5034 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14356 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47988 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6797 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10279 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

donny-dont commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -340,7 +340,7 @@ void GPUConnectionToWebProcess::didClose(IPC::Connection& connection)
+#if USE(GRAPHICS_LAYER_WC) && ENABLE(WEBGL)
     remoteGraphicsContextGLStreamWorkQueue().dispatch([webProcessIdentifier = m_webProcessIdentifier] {
         WCContentBufferManager::singleton().removeAllContentBuffersForProcess(webProcessIdentifier);
     });
```
This whole area feels incorrect.

fujii replied:
I think this is correct. WCContentBufferIdentifier is used only for WebGL now.
You can add ENABLE(WEBGL) for Source/WebKit/Shared/wc/WCContentBufferIdentifier.h too.
I don't know how to add media support GraphicLayerWC now. Anyway, it's a future plan. It's OK for now.

donny-dont commented on Source/WebKit/Shared/WebPageCreationParameters.h:
```diff
@@ -187,7 +187,7 @@ struct WebPageCreationParameters {
 #endif
-#if PLATFORM(WIN)
+#if PLATFORM(WIN) || USE(GRAPHICS_LAYER_WC)
     uint64_t nativeWindowHandle;
```
Is `nativeWindowHandle` required for WC?

I don't think all the plumbing is there for PlayStation if so.

fujii replied:
Yes. `nativeWindowHandle` is used for `GLContext::createContextForWindow`.
https://github.com/WebKit/WebKit/blob/12f417282701d97084722aa8cad975f458fd66e4/Source/WebKit/GPUProcess/graphics/wc/WCSceneContext.cpp#L38
I have a plan to pass the `nativeWindowHandle` from UI process to GPU process directly in the future.
For the security, I don't want to tell web process the `nativeWindowHandle`.

donny-dont commented on Source/WebKit/UIProcess/playstation/PageClientImpl.cpp:
```diff
@@ -191,6 +201,13 @@ RefPtr<WebPopupMenuProxy> PageClientImpl::createPopupMenuProxy(WebPageProxy&  pa
+#if USE(GRAPHICS_LAYER_WC)
+bool PageClientImpl::usesOffscreenRendering() const
+{
+    return false;
```
This relevant for PlayStation?

fujii replied:
It should be same with WinCairo. Or, should share the implementation.
usesOffscreenRendering is used for pixel dump of AC mode for WebKitTestRunner.

donny-dont replied:
WebKitTestRunner isn't enabled for PS in main so I'm going to defer an implementation at this point

donny-dont reviewed (commented):
@fujii WIP with a couple questions for you.

fujii commented on Source/WebKit/GPUProcess/graphics/wc/RemoteWCLayerTreeHost.cpp:
```diff
@@ -44,14 +44,16 @@ std::unique_ptr<RemoteWCLayerTreeHost> RemoteWCLayerTreeHost::create(GPUConnecti
 
 RemoteWCLayerTreeHost::RemoteWCLayerTreeHost(GPUConnectionToWebProcess& connectionToWebProcess, WebKit::WCLayerTreeHostIdentifier identifier, uint64_t nativeWindow, bool usesOffscreenRendering)
-    : m_connectionToWebProcess(connectionToWebProcess)
+    : m_workQueue(IPC::StreamConnectionWorkQueue::create("RemoteWCLayerTreeHost work queue"))
```
ANGLE D3D renderer isn't thread-safe. WebGL and TextureMaper should be run in the same thread of GPU process. Don't create a new work queue.

fujii commented on Source/WebKit/UIProcess/playstation/PageClientImpl.cpp:
```diff
@@ -40,6 +44,12 @@ PageClientImpl::PageClientImpl(PlayStationWebView& view)
+#if USE(GRAPHICS_LAYER_WC)
+    if (m_view.page()->preferences().useGPUProcessForWebGLEnabled() ||
+        m_view.page()->preferences().useGPUProcessForDOMRenderingEnabled() ||
+        m_view.page()->preferences().useGPUProcessForDOMRenderingEnabled())
```
useGPUProcessForDOMRenderingEnabled is duplicated.,

fujii commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -2141,7 +2141,7 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
 #endif
 
-#if PLATFORM(WIN)
+#if PLATFORM(WIN) || USE(GRAPHICS_LAYER_WC)
```
Why don't you remove "PLATFORM(WIN) ||"?

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -51,7 +51,9 @@ PageClientImpl::PageClientImpl(WebView& view)
-    if (m_view.page()->preferences().useGPUProcessForWebGLEnabled())
+    if (m_view.page()->preferences().useGPUProcessForWebGLEnabled() ||
+        m_view.page()->preferences().useGPUProcessForDOMRenderingEnabled() ||
+        m_view.page()->preferences().useGPUProcessForDOMRenderingEnabled())
```
WinCairo uses DrawingAreaProxyWC if useGPUProcessForWebGLEnabled.
PlayStation doesn't support WebGL. It should use DrawingAreaProxyWC unconditinally.
>if USE(GRAPHICS_LAYER_WC)
>if ENABLE(WEBGL)
>    bool usesWC = m_view.page()->preferences().useGPUProcessForWebGLEnabled();
>#else
>   bool usesWC = false
>#endif
>   if (usesWC)

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -51,7 +51,9 @@ PageClientImpl::PageClientImpl(WebView& view)
-    if (m_view.page()->preferences().useGPUProcessForWebGLEnabled())
+    if (m_view.page()->preferences().useGPUProcessForDOMRenderingEnabled()
+        || m_view.page()->preferences().useGPUProcessForCanvasRenderingEnabled()
+        || m_view.page()->preferences().useGPUProcessForWebGLEnabled())
```
Revert this part. If WebGL is running in web process, the compositor in GPU process can't get the output texture. WebGL and the compositor should be in the same process for WinCairo.

fujii reviewed (approved):
WinCairo EWS is down. Please check it before landing.
LGTM.

donny-dont committed:
WC shouldn't require ENABLE_WEBGL
https://bugs.webkit.org/show_bug.cgi?id=245188

Reviewed by Fujii Hironori.

Make it so `USE(GRAPHICS_LAYER_WC)` can work without `ENABLE(WEBGL)`.
Guard `ENABLE(WEBGL)` usage and add a single `StreamConnectionWorkQueue`
to submit commands to. This is done due to thread safety issues with
ANGLE.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
* Source/WebKit/GPUProcess/graphics/wc/RemoteWCLayerTreeHost.cpp:
* Source/WebKit/UIProcess/playstation/PageClientImpl.cpp:
* Source/WebKit/UIProcess/playstation/PageClientImpl.h:
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:

Canonical link: https://commits.webkit.org/259825@main
